### PR TITLE
Tezos 8.0

### PR DESCRIPTION
**Pull request by pirbo**
This is once again an hijack of the CI (But if nothing catastrophic is raised by it the real release is for tomorrow the latest)

### CODE REVIEW

File: packages/tezos-protocol-environment-packer/tezos-protocol-environment-packer.8.0~rc2/opam [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+depends: [
+  "dune" { >= "2.0" }
+]
+build: [
+  ["dune" "build" "-p" name "-j" jobs]
+  ["mv" "src/lib_protocol_environment/%{name}%.install" "./"]
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
+]
+synopsis: "Tezos: sigs/structs packer for economic protocol environment"
+
```
Comment (avsm) [inline packages/tezos-protocol-environment-packer/tezos-protocol-environment-packer.8.0~rc2/opam:17]: This one needs to be ocaml-version >= 4.03 at least due to String.capitalize_ascii

File: packages/tezos-stdlib/tezos-stdlib.8.0~rc2/opam [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+depends: [
+  "dune" { >= "2.0" }
+  "ocaml" { >= "4.8.0" }
+  "hex"
+  "lwt"
+  "zarith"
+  "bigstring" { with-test }
+  "lwt_log" { with-test }
+  "alcotest" { with-test & >= "1.1.0" }
+  "alcotest-lwt" { with-test & >= "1.1.0" }
```
Comment (avsm) [inline packages/tezos-stdlib/tezos-stdlib.8.0~rc2/opam:17]: tezos-stdlib needs a `crowbar` {with-test} dependency

File: packages/tezos-lwt-result-stdlib/tezos-lwt-result-stdlib.8.0~rc2/opam [hunk 3]
```diff
@@ -0,0 +1,27 @@
+opam-version: "2.0"
+maintainer: "contact@tezos.com"
+authors: [ "Tezos devteam" ]
+homepage: "https://www.tezos.com/"
+bug-reports: "https://gitlab.com/tezos/tezos/issues"
+dev-repo: "git+https://gitlab.com/tezos/tezos.git"
+license: "MIT"
+depends: [
+  "dune" { >= "2.0" }
```
Comment (avsm) [inline packages/tezos-lwt-result-stdlib/tezos-lwt-result-stdlib.8.0~rc2/opam:9]: All these packages need a dependency on `ocaml` as well. In this particular library, that needs to be <4.12.0 due to this:

```
#=== ERROR while compiling tezos-lwt-result-stdlib.8.0~rc2 ====================#
# context              2.0.7 | linux/x86_64 | ocaml-variants.4.12.0+trunk | file:///src
# path                 ~/.opam/4.12/.opam-switch/build/tezos-lwt-result-stdlib.8.0~rc2
# command              ~/.opam/4.12/bin/dune build -p tezos-lwt-result-stdlib -j 47
# exit-code            1
# env-file             ~/.opam/log/tezos-lwt-result-stdlib-1-531662.env
# output-file          ~/.opam/log/tezos-lwt-result-stdlib-1-531662.out
### output ###
#       ocamlc src/lib_lwt_result_stdlib/sigs/.sigs.objs/byte/sigs__Hashtbl.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib_lwt_result_stdlib/sigs/.sigs.objs/byte -I /home/opam/.opam/4.12/lib/bytes -I /home/opam/.opam/4.12/lib/lwt -I /home/opam/.opam/4.12/lib/result -I /home/opam/.opam/4.12/lib/seq -no-alias-deps -open Sigs -o src/lib_lwt_result_stdlib/sigs/.sigs.objs/byte/sigs__Hashtbl.cmo -c -impl src/lib_lwt_result_stdlib/sigs/hashtbl.ml)
# File "src/lib_lwt_result_stdlib/sigs/hashtbl.ml", line 36, characters 18-75:
# 36 |   module Legacy : Stdlib.Hashtbl.S with type key = key and type 'a t = 'a t
#                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: In this `with' constraint, the new definition of t
#        does not match its original definition in the constrained signature:
#        Type declarations do not match:
#          type 'a t = 'a t
#        is not included in
#          type !'a t
#        Their variances do not agree.
#        File "hashtbl.mli", line 335, characters 4-14: Expected declaration
#        File "src/lib_lwt_result_stdlib/sigs/hashtbl.ml", line 36, characters 59-75:
#          Actual declaration
```

File: packages/tezos-base/tezos-base.8.0~rc2/opam [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+depends: [
+  "dune" { >= "2.0" }
+  "tezos-crypto" { = version }
+  "tezos-micheline" { = version }
+  "ptime" { >= "0.8.4" }
+  "ezjsonm" { >= "0.5.0" }
+  "ipaddr" {>= "5.0.0" & < "6.0.0"}
+  "crowbar" { with-test }
+]
+build: [
```
Comment (avsm) [inline packages/tezos-base/tezos-base.8.0~rc2/opam:17]: tezos-base needs to marked `available: "arch" != "arm32"` due to

```
#=== ERROR while compiling tezos-base.8.0~rc2 =================================#
# context              2.1.0~beta4 | linux/arm32 | ocaml-base-compiler.4.11.1 | pinned(https://gitlab.com/nomadic-labs/tezos/-/archive/d88aaf75c7c6de5358de75005791f9d3f049c3ba/tezos-d88aaf75c7c6de5358de75005791f9d3f049c3ba.tar.bz2)
# path                 ~/.opam/4.11/.opam-switch/build/tezos-base.8.0~rc2
# command              ~/.opam/4.11/bin/dune runtest -p tezos-base -j 31
# exit-code            1
# env-file             ~/.opam/log/tezos-base-7-f8c573.env
# output-file          ~/.opam/log/tezos-base-7-f8c573.out
### output ###
#       ocamlc src/lib_base/test/.test_time.eobjs/byte/dune__exe__Test_time.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.11/bin/ocamlc.opt -w -40 -open Tezos_base -g -bin-annot -I src/lib_base/test/.test_time.eobjs/byte -I /home/opam/.opam/4.11/lib/afl-persistent -I /home/opam/.opam/4.11/lib/angstrom -I /home/opam/.opam/4.11/lib/astring -I /home/opam/.opam/4.11/lib/bigarray-compat -I /home/opam/.opam/4.11/lib/bigstring -I /home/opam/.opam/4.11/lib/bigstringaf -I /home/opam/.opam/4.11/lib/bls12-381 -I /home/opam/.opam/4.11/lib/bytes -I /home/opam/.opam/4.11/lib/cmdliner -I /home/opam/.opam/4.11/lib/crowbar -I /home/opam/.opam/4.11/lib/cstruct -I /home/opam/.opam/4.11/lib/ctypes -I /home/opam/.opam/4.11/lib/data-encoding -I /home/opam/.opam/4.11/lib/domain-name -I /home/opam/.opam/4.11/lib/ezjsonm -I /home/opam/.opam/4.11/lib/ff -I /home/opam/.opam/4.11/lib/fmt -I /home/opam/.opam/4.11/lib/hacl-star -I /home/opam/.opam/4.11/lib/hacl-star-raw -I /home/opam/.opam/4.11/lib/hex -I /home/opam/.opam/4.11/lib/integers -I /home/opam/.opam/4.11/lib/ipaddr -I /home/opam/.opam/4.11/lib/json-data-encoding -I /home/opam/.opam/4.11/lib/json-data-encoding-bson -I /home/opam/.opam/4.11/lib/jsonm -I /home/opam/.opam/4.11/lib/lwt -I /home/opam/.opam/4.11/lib/lwt-canceler -I /home/opam/.opam/4.11/lib/lwt_log/core -I /home/opam/.opam/4.11/lib/macaddr -I /home/opam/.opam/4.11/lib/ocaml/threads -I /home/opam/.opam/4.11/lib/ocplib-endian -I /home/opam/.opam/4.11/lib/ptime -I /home/opam/.opam/4.11/lib/re -I /home/opam/.opam/4.11/lib/resto -I /home/opam/.opam/4.11/lib/resto-directory -I /home/opam/.opam/4.11/lib/result -I /home/opam/.opam/4.11/lib/ringo -I /home/opam/.opam/4.11/lib/secp256k1-internal -I /home/opam/.opam/4.11/lib/seq -I /home/opam/.opam/4.11/lib/sexplib0 -I /home/opam/.opam/4.11/lib/stdlib-shims -I /home/opam/.opam/4.11/lib/stringext -I /home/opam/.opam/4.11/lib/tezos-clic -I /home/opam/.opam/4.11/lib/tezos-crypto -I /home/opam/.opam/4.11/lib/tezos-error-monad -I /home/opam/.opam/4.11/lib/tezos-event-logging -I /home/opam/.opam/4.11/lib/tezos-lwt-result-stdlib -I /home/opam/.opam/4.11/lib/tezos-lwt-result-stdlib/functors -I /home/opam/.opam/4.11/lib/tezos-lwt-result-stdlib/lib -I /home/opam/.opam/4.11/lib/tezos-lwt-result-stdlib/sigs -I /home/opam/.opam/4.11/lib/tezos-micheline -I /home/opam/.opam/4.11/lib/tezos-rpc -I /home/opam/.opam/4.11/lib/tezos-stdlib -I /home/opam/.opam/4.11/lib/uchar -I /home/opam/.opam/4.11/lib/uecc -I /home/opam/.opam/4.11/lib/uri -I /home/opam/.opam/4.11/lib/uutf -I /home/opam/.opam/4.11/lib/zarith -I src/lib_base/.tezos_base.objs/byte -I src/lib_base/.tezos_base.objs/native -no-alias-deps -open Dune__exe -o src/lib_base/test/.test_time.eobjs/byte/dune__exe__Test_time.cmo -c -impl src/lib_base/test/test_time.ml)
# File "src/lib_base/test/test_time.ml", line 119, characters 36-58:
# 119 |       [range (day_range + 2); range 86_400_000_000_000_000]
#                                           ^^^^^^^^^^^^^^^^^^^^^^
# Error: Integer literal exceeds the range of representable integers of type int
# test_p2p_addr alias src/lib_base/test/runtest_p2p_addr
# Base.P2p_addr.to-string-from-string: PASS
# 
```

**Review by avsm — CHANGES_REQUESTED** [review]
Generally looks pretty good. There is a problem running tests with a missing kremlin header; if you could fix that then the CI should start to go fairly green.  Other comments left inline.

**Comment by pirbo** [comment]
Needs (and cherry-pick for now) #17866 Except from that, should be ready!

**Comment by avsm** [comment]
A lot of these packages are still missing dependencies on "ocaml". Is that intentional? It'll be pulled in transitively, but it would be a good idea to depend on it explicitly and also to put a base version in there that you are expecting to support that library on. (it'll let the solver give better error messages).

**Comment by avsm** [comment]
Looks like #17866 wasn't cherry picked here. Could you rebase against master where it has been merged, so that's picked up for this PR?

```
    - tezos-embedded-protocol-002-PsYLVpVv >= 8.0 -> tezos-protocol-updater >= 8.0 -> tezos-protocol-compiler >= 8.0 -> tezos-protocol-environment >= 8.0 -> tezos-sapling -> tezos-crypto >= 8.0 -> bls12-381 < 0.3.14
```

**Comment by pirbo** [comment]
For bls, there were indeed a mistake I fixed.

For OCaml version constraint, yes, I counted on the fact that `tezos-stdlib` and `tezos-protocol-compiler` would transitively constraint everything else but I understand that being more explicit would be better

Finally for the compatibility with 32bits, There is an ongoing discussion about explicitly forbidding them (because their support has a huge consequence on the efficiency of the interpreter of smart contract). I'll do something cleaner when we reach a conclusion. (By the way, is there a cleaner way to express "not compatible with 32bits" in opam than `available: [ arch != "x86_32" & arch != "arm32" & arch != "ppc32" ]` ?)

**Comment by avsm** [comment]
That `available` command is the best we can do for now. We'll fix things up properly in a future revision of the language, but the only other 32-bit architecture on the horizon is a possible riscv32 when hardware becomes available -- I dont think this'll be a problem in practise :-)

**Comment by avsm** [comment]
This looks like a potential blocker:

```
#=== ERROR while compiling tezos-mockup-registration.8.0 ======================#
# context              2.0.7 | linux/x86_64 | ocaml-base-compiler.4.09.1 | file:///src
# path                 ~/.opam/4.09/.opam-switch/build/tezos-mockup-registration.8.0
# command              ~/.opam/4.09/bin/dune build -p tezos-mockup-registration -j 47
# exit-code            1
# env-file             ~/.opam/log/tezos-mockup-registration-1-74f863.env
# output-file          ~/.opam/log/tezos-mockup-registration-1-74f863.out
### output ###
# File "src/lib_mockup/dune", line 8, characters 3-23:
# 8 |    tezos-shell-services
#        ^^^^^^^^^^^^^^^^^^^^
# Error: Library "tezos-shell-services" not found.
# Hint: try:
#   dune external-lib-deps --missing -p tezos-mockup-registration -j 47 @install

```

**Comment by pirbo** [comment]
I'm sorry I completely mess-up a rebase... It should be better now. 
There are still a lot of tests "skipped" with fishy error messages like
```
[ERROR] No solution for tezos-mockup-commands.8.0: The actions to process have cyclic dependencies:
          - install irmin.2.2.0 -> install ppx_irmin.2.2.0 -> install irmin.2.2.0
```
but I don't think it is related to this PR (or at least I don't reproduce locally)

**Comment by kit-ty-kate** [comment]
Any change you know where this comes from? This test failure seems to happen randomly
```
#=== ERROR while compiling tezos-stdlib-unix.8.0 ==============================#
# context              2.0.7 | linux/x86_64 | ocaml-base-compiler.4.08.1 | pinned(https://gitlab.com/nomadic-labs/tezos/-/archive/v8.0/tezos-v8.0.tar.bz2)
# path                 ~/.opam/4.08/.opam-switch/build/tezos-stdlib-unix.8.0
# command              ~/.opam/4.08/bin/dune runtest -p tezos-stdlib-unix -j 47
# exit-code            1
# env-file             ~/.opam/log/tezos-stdlib-unix-1-d8a4ae.env
# output-file          ~/.opam/log/tezos-stdlib-unix-1-d8a4ae.out
### output ###
# test_lwt_exit_main_is_clean_up alias src/lib_stdlib_unix/test/runtest_test_lwt_exit_main_is_clean_up (exit 2)
# (cd _build/default/src/lib_stdlib_unix/test && ./test_lwt_exit_main_is_clean_up.exe)
# Fatal error: exception File "src/lib_stdlib_unix/test/test_lwt_exit_main_is_clean_up.ml", line 102, characters 6-12: Assertion failed
```
```
#=== ERROR while compiling tezos-stdlib-unix.8.0 ==============================#
# context              2.0.7 | linux/x86_64 | ocaml-base-compiler.4.11.1 | pinned(https://gitlab.com/nomadic-labs/tezos/-/archive/v8.0/tezos-v8.0.tar.bz2)
# path                 ~/.opam/4.11/.opam-switch/build/tezos-stdlib-unix.8.0
# command              ~/.opam/4.11/bin/dune runtest -p tezos-stdlib-unix -j 31
# exit-code            1
# env-file             ~/.opam/log/tezos-stdlib-unix-1-d8a4ae.env
# output-file          ~/.opam/log/tezos-stdlib-unix-1-d8a4ae.out
### output ###
# test_lwt_exit_main_is_clean_up alias src/lib_stdlib_unix/test/runtest_test_lwt_exit_main_is_clean_up (exit 2)
# (cd _build/default/src/lib_stdlib_unix/test && ./test_lwt_exit_main_is_clean_up.exe)
# Fatal error: exception File "src/lib_stdlib_unix/test/test_lwt_exit_main_is_clean_up.ml", line 102, characters 6-12: Assertion failed
```

**Comment by pirbo** [comment]
I asked @raphael-proust for its opinion. There might be something fishy under the hood but as the rest of tezos is only compatible with OCaml 4.09 and OCaml 4.10 and as I "retried" a few time and never observed failure there, we may live with that for this time and leave the fix for version 8.1 :-|

**Comment by kit-ty-kate** [comment]
This also happens with OCaml 4.11.1 as shown above. I might have to disable the tests for this package specifically in the future if this failure shows up again.

Thanks.

**Pull request merged**